syntax = "proto2";

package detection.protos;

message Preprocessor {
  repeated PreprocessingStep data_augmentation_options = 2;
}

message PreprocessingStep {
  oneof preprocessing_step_oneof {
    RandomHorizontalFlip random_horizontal_flip = 2;
    RandomCropImage random_crop_image = 12;
    RandomPadImage random_pad_image = 13;
    SSDRandomCrop ssd_random_crop = 21;
    ResizeImage resize_image = 19;
  }
}

message RandomHorizontalFlip {
}

message RandomPadImage {
  optional float pad_ratio = 1 [default = 1.0];

  optional float keep_prob = 2 [default = 1.0];
}

message RandomCropImage {
  optional float min_object_covered = 1 [default = 1.0];

  optional float min_aspect_ratio = 2 [default = 0.75];

  optional float max_aspect_ratio = 3 [default = 1.33];

  optional float min_area = 4 [default = 0.1];

  optional float max_area = 5 [default = 1.0];

  optional float overlap_thresh = 6 [default = 0.3];

  optional float prob_to_keep_original = 7 [default = 0.0];
}


message SSDRandomCrop {
  repeated RandomCropImage operations = 1;
}

message ResizeImage {
  optional int32 new_height = 1;
  optional int32 new_width = 2;
  enum Method {
    AREA = 1;
    BICUBIC = 2;
    BILINEAR = 3;
    NEAREST_NEIGHBOR = 4;
  }
  optional Method method = 3 [default = BILINEAR];
}


message Normalizer {
  enum Type {
    RANGE = 1;
    SUBTRACT_MEAN = 2;  
  }

  optional Type type = 1;

  optional float low = 2 [default = -1.0];

  optional float high = 3 [default = 1.0];

  optional float r_mean = 4 [default = 123.68];

  optional float g_mean = 5 [default = 116.779];

  optional float b_mean = 6 [default = 103.939];
}

